File: claude/web/src/lib/arcade/storm.ts
/**
 * Storm — a deterministic SteamBench arcade *multi-object temporal-vision* env.
 *
 * TypeScript mirror of `harness/steambench_harness/envs/storm.py` (and the
 * headless twin of the pixel runtime's StormGame). Blocks rain down, each at its
 * OWN speed, so you must track several at once.
 *
 * Determinism note: integer positions/velocities in Python's exact order. Each
 * spawn draws TWO values in a fixed order — the column `rng.randrange(W - BW + 1)`,
 * then the fall speed `_VY_MIN + rng.randrange(_VY_SPAN)`. Spawns fire on a fixed
 * tick cadence (`steps % SPAWN_EVERY === 0`), so `(seed, actions)` traces replay
 * identically — human, Python agent, or the vision agent reading frames.
 */

import { ActionSpace, Env } from "./base";
import type { AchievementSpec, Action, Observation } from "./types";

const W = 168;
const H = 120;
const PW = 22;
const PH = 8;
const BW = 12;
const BH = 12;
const PLAYER_Y = H - PH - 3; // 109
const PSPEED = 7;
const SPAWN_EVERY = 8;
const VY_MIN = 3;
const VY_SPAN = 5;
const ACTIONS = ["left", "stay", "right"] as const;

export class Storm extends Env {
  static env_id = "arcade/storm";
  static appid = 9000012;
  static displayName = "Storm";
  static description =
    "Dodge the falling blocks — but every block falls at its own speed, so you " +
    "must track several at once and read each one's velocity to survive. " +
    "Multi-object temporal perception: the frontier a vision agent plays from raw frames.";
  static verify_mode = "replay" as const;
  static action_space = new ActionSpace([...ACTIONS]);
  static achievements: AchievementSpec[] = [
    { id: "survive_50", name: "Drizzle", description: "Survive 50 ticks.", rarity_hint: 0.6 },
    { id: "survive_150", name: "Downpour", description: "Survive 150 ticks.", rarity_hint: 0.3 },
    { id: "survive_300", name: "Squall", description: "Survive 300 ticks.", rarity_hint: 0.12 },
    { id: "survive_600", name: "Tempest", description: "Survive 600 ticks.", rarity_hint: 0.035 },
    { id: "survive_1200", name: "Eye of the Storm", description: "Survive 1200 ticks.", rarity_hint: 0.006 },
  ];
  private static LADDER: [number, string][] = [
    [50, "survive_50"], [150, "survive_150"], [300, "survive_300"],
    [600, "survive_600"], [1200, "survive_1200"],
  ];

  px: number;
  blocks: [number, number, number][]; // [x, y, vy]
  alive: boolean;

  constructor() {
    super();
    this.px = Math.trunc((W - PW) / 2);
    this.blocks = [];
    this.alive = true;
  }

  reset(seed = 0): Observation {
    this.begin(seed);
    this.px = Math.trunc((W - PW) / 2);
    this.blocks = [];
    this.alive = true;
    return this.observe(0.0);
  }

  step(action: Action): Observation {
    if (this.done) return this.observe(0.0);
    const name = this.actionSpace.name(action);
    this.steps += 1;

    if (name === "left") this.px = Math.max(0, this.px - PSPEED);
    else if (name === "right") this.px = Math.min(W - PW, this.px + PSPEED);

    if (this.steps % SPAWN_EVERY === 0) {
      const x = this.rng.randrange(W - BW + 1);
      const vy = VY_MIN + this.rng.randrange(VY_SPAN);
      this.blocks.push([x, -BH, vy]);
    }
    for (const b of this.blocks) b[1] += b[2];
    this.blocks = this.blocks.filter((b) => b[1] < H);

    let hit = false;
    for (const [bx, by] of this.blocks) {
      if (bx < this.px + PW && bx + BW > this.px && by < PLAYER_Y + PH && by + BH > PLAYER_Y) {
        hit = true;
        break;
      }
    }

    let reward: number;
    if (hit) {
      this.alive = false;
      this.done = true;
      reward = -1.0;
    } else {
      this.score += 1;
      reward = 1.0;
      for (const [thresh, aid] of Storm.LADDER) {
        if (this.score >= thresh) this.unlock(aid);
      }
    }
    return this.observe(reward);
  }

  legalActions(): string[] {
    return [...ACTIONS];
  }

  render(): string {
    return `score=${Math.trunc(this.score)} px=${this.px} blocks=${this.blocks.length} ${this.alive ? "ALIVE" : "DEAD"}`;
  }

  private observe(reward: number): Observation {
    return {
      step: this.steps,
      state: {
        paddle_x: this.px, paddle_y: PLAYER_Y, paddle_w: PW, paddle_h: PH, paddle_speed: PSPEED,
        block_w: BW, block_h: BH, width: W, height: H,
        blocks: this.blocks.map(([x, y, vy]) => ({ x, y, vy })),
        score: Math.trunc(this.score), alive: this.alive,
      },
      text: this.render(),
      frame: null,
      legal_actions: this.legalActions(),
      score: this.score,
      done: this.done,
      reward: reward,
      info: { score: Math.trunc(this.score), alive: this.alive },
    };
  }
}

File: claude/harness/steambench_harness/envs/storm.py
"""Storm — a deterministic SteamBench arcade *multi-object temporal-vision* env.

Blocks rain down, but unlike Dodger every block falls at its OWN speed — fast
ones reach you sooner than slow ones spawned earlier. To survive you must track
*several* objects at once and know each one's velocity, not just its position.
That multi-object temporal tracking (match blocks across frames, infer each
one's speed) is the next vision capability after Volley's single ball, and it is
what the pixel runtime's CV agent demonstrates here.

Integer positions/velocities → bit-identical across Python and TypeScript; the
only randomness is each block's spawn column and fall speed (two draws per
spawn, in order), so ``(seed, actions)`` traces replay exactly.
"""

from __future__ import annotations

from typing import Union

from ..protocol import AchievementSpec, ActionSpace, Env, Observation, VerifyMode
from ..registry import register

# Shared with runtime/pixel_game.py StormGame — keep in lock-step (parity test).
W, H = 168, 120
PW, PH = 22, 8
BW, BH = 12, 12
PLAYER_Y = H - PH - 3   # 109
PSPEED = 7
SPAWN_EVERY = 8         # spawn a block every N ticks
_VY_MIN, _VY_SPAN = 3, 5   # fall speed in [3, 7] (one rng draw: 3 + randrange(5))
_ACTIONS = ["left", "stay", "right"]


@register
class Storm(Env):
    env_id = "arcade/storm"
    appid = 9000012
    name = "Storm"
    description = (
        "Dodge the falling blocks — but every block falls at its own speed, so "
        "you must track several at once and read each one's velocity to survive. "
        "Multi-object temporal perception: the frontier a vision agent plays from "
        "raw frames."
    )
    verify_mode = VerifyMode.REPLAY
    action_space = ActionSpace(_ACTIONS)
    achievements = [
        AchievementSpec("survive_50", "Drizzle", "Survive 50 ticks.", 0.6),
        AchievementSpec("survive_150", "Downpour", "Survive 150 ticks.", 0.3),
        AchievementSpec("survive_300", "Squall", "Survive 300 ticks.", 0.12),
        AchievementSpec("survive_600", "Tempest", "Survive 600 ticks.", 0.035),
        AchievementSpec("survive_1200", "Eye of the Storm", "Survive 1200 ticks.", 0.006),
    ]
    _LADDER = [(50, "survive_50"), (150, "survive_150"), (300, "survive_300"),
               (600, "survive_600"), (1200, "survive_1200")]

    def __init__(self) -> None:
        super().__init__()
        self.px = (W - PW) // 2
        self.blocks: list[list[int]] = []   # [x, y, vy]
        self.alive = True

    def reset(self, seed: int = 0) -> Observation:
        self._begin(seed)
        self.px = (W - PW) // 2
        self.blocks = []
        self.alive = True
        return self._observe(0.0)

    def step(self, action: Union[int, str]) -> Observation:
        if self.done:
            return self._observe(0.0)
        name = self.action_space.name(action)
        self.steps += 1

        if name == "left":
            self.px = max(0, self.px - PSPEED)
        elif name == "right":
            self.px = min(W - PW, self.px + PSPEED)

        # Spawn on a fixed tick cadence (two draws: column, then fall speed).
        if self.steps % SPAWN_EVERY == 0:
            x = self.rng.randrange(W - BW + 1)
            vy = _VY_MIN + self.rng.randrange(_VY_SPAN)
            self.blocks.append([x, -BH, vy])
        for blk in self.blocks:
            blk[1] += blk[2]
        self.blocks = [b for b in self.blocks if b[1] < H]

        hit = False
        for bx, by, _vy in self.blocks:
            if (bx < self.px + PW and bx + BW > self.px and
                    by < PLAYER_Y + PH and by + BH > PLAYER_Y):
                hit = True
                break

        if hit:
            self.alive = False
            self.done = True
            reward = -1.0
        else:
            self.score += 1
            reward = 1.0
            for thresh, aid in self._LADDER:
                if self.score >= thresh:
                    self._unlock(aid)
        return self._observe(reward)

    def legal_actions(self) -> list[str]:
        return list(_ACTIONS)

    def render(self) -> str:
        return (f"score={int(self.score)} px={self.px} blocks={len(self.blocks)} "
                f"{'ALIVE' if self.alive else 'DEAD'}")

    def _observe(self, reward: float) -> Observation:
        return Observation(
            step=self.steps,
            state={
                "paddle_x": self.px,
                "paddle_y": PLAYER_Y,
                "paddle_w": PW,
                "paddle_h": PH,
                "paddle_speed": PSPEED,
                "block_w": BW,
                "block_h": BH,
                "width": W,
                "height": H,
                "blocks": [{"x": bx, "y": by, "vy": vy} for bx, by, vy in self.blocks],
                "score": int(self.score),
                "alive": self.alive,
            },
            text=self.render(),
            legal_actions=self.legal_actions(),
            score=float(self.score),
            done=self.done,
            reward=float(reward),
            info={"score": int(self.score), "alive": self.alive},
        )
